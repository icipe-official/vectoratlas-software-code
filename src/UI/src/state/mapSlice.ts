import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { fetchApiJson } from '../api/api';

export interface MapState {
  map_styles: {
    layers: {
      name: string;
      fillColor: number[];
      strokeColor: number[];
      strokeWidth: number;
      zIndex: number;
    }[];
  };

  map_overlays: {
    name: string;
    source: string;
    sourceType: string;
    layers: { name: string }[];
  }[];
}

export const initialState: MapState = {
  map_styles: { layers: [] },
  map_overlays: [],
};

export const getMapStyles = createAsyncThunk('map/getMapStyles', async () => {
  const mapStyles = await fetchApiJson('config/map-styles');
  return mapStyles;
});

export const getTileServerOverlays = createAsyncThunk(
  'map/getTileServerOverlays',
  async () => {
    const tileServerOverlays = await fetchApiJson(
      'config/tile-server-overlays'
    );
    return tileServerOverlays;
  }
);

export const mapSlice = createSlice({
  name: 'map',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getMapStyles.fulfilled, (state, action) => {
        state.map_styles = action.payload;
      })
      .addCase(getTileServerOverlays.fulfilled, (state, action) => {
        state.map_overlays = action.payload;
      });
  },
});

export default mapSlice.reducer;
