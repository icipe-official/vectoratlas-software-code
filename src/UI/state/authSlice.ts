import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import * as njwt from 'njwt';
import { fetchAuth } from '../api/api';

export interface AuthState {
  roles: String[];
  token: String;
}

export const initialState: AuthState = {
  roles: [],
  token: '',
};

export const getUserInfo = createAsyncThunk('auth/getUserInfo', async () => {
  const token = await fetchAuth();
  const verifiedToken: any = njwt.verify(
    token,
    process.env.NEXT_PUBLIC_TOKEN_KEY
  );
  return {
    token: token,
    roles: verifiedToken?.body.scope.split(','),
  };
});

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getUserInfo.fulfilled, (state, action: any) => {
      state.roles = action.payload.roles;
      state.token = action.payload.token;
    });
  },
});

export default authSlice.reducer;
