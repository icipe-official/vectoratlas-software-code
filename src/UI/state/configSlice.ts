import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import {
  fetchLocalVersion,
  fetchApiVersion,
  fetchFeatureFlags,
} from '../api/api';

export interface ConfigState {
  version_ui: string;
  version_api: string;
  feature_flags_status: string;
  feature_flags: {
    flag: string;
    on: boolean;
  }[];
}

export const initialState: ConfigState = {
  version_ui: 'local_ui',
  version_api: 'local_api',
  feature_flags: [],
  feature_flags_status: '',
};

export const getUiVersion = createAsyncThunk(
  'config/getUiVersion',
  async () => {
    const version = await fetchLocalVersion();
    return version;
  }
);

export const getApiVersion = createAsyncThunk(
  'config/getApiVersion',
  async () => {
    const version = await fetchApiVersion();
    return version;
  }
);

export const getFeatureFlags = createAsyncThunk(
  'config/getFeatureFlags',
  async () => {
    const featureFlags = await fetchFeatureFlags();
    return featureFlags;
  }
);

export const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getUiVersion.pending, (state) => {
        state.version_ui = 'loading';
      })
      .addCase(getUiVersion.rejected, (state) => {
        state.version_ui = 'error';
      })
      .addCase(getUiVersion.fulfilled, (state, action) => {
        state.version_ui = action.payload;
      })
      .addCase(getApiVersion.pending, (state) => {
        state.version_api = 'loading';
      })
      .addCase(getApiVersion.rejected, (state) => {
        state.version_api = 'error';
      })
      .addCase(getApiVersion.fulfilled, (state, action) => {
        state.version_api = action.payload;
      })
      .addCase(getFeatureFlags.pending, (state) => {
        state.feature_flags_status = 'loading';
      })
      .addCase(getFeatureFlags.rejected, (state) => {
        state.feature_flags_status = 'error';
      })
      .addCase(getFeatureFlags.fulfilled, (state, action) => {
        state.feature_flags_status = 'success';
        state.feature_flags = action.payload;
      });
  },
});

export default configSlice.reducer;
