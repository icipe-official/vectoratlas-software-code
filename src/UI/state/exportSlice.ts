// import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
// import { fetchApiJson } from '../api/api';

// export const getAll = createAsyncThunk('export/getAllOccurrenceData', async () => {
//     const allOccurrenceData = await fetchApiJson('export/stream-all-file');
//     return allOccurrenceData;
//   }
// );

// export const initialState: any = []

// export const mapSlice = createSlice({
//   name: 'map',
//   initialState
//   reducers: {}
    
//   extraReducers: (builder) => {
//     builder
//       .addCase(getMapStyles.fulfilled, (state, action) => {
//         state.map_styles = action.payload;
//       })
//       .addCase(getTileServerOverlays.fulfilled, (state, action) => {
//         state.map_overlays = action.payload;
//       });

//   },
// });

// export const {
  
// } = mapSlice.actions;

// export default mapSlice.reducer;
