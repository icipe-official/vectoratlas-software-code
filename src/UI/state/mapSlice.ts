import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import {
  fetchGraphQlData,
  fetchMapStyles,
  fetchTileServerOverlays,
} from '../api/api';
import { locationsQuery } from '../api/queries';
export interface MapState {
  map_styles: {
    layers: {
      name: string;
      fillColor: number[];
      strokeColor: number[];
      strokeWidth: number;
      zIndex: number;
    }[];
  };

  map_overlays: {
    name: string;
    source: string;
    sourceType: string;
    layers: { name: string }[];
  }[];

  occurrence_data: {
    items: [{}];
    total: number;
    hasMore: boolean;
  }[];

  map_drawer: {
    open: boolean;
    overlays: boolean;
    baseMap: boolean;
  };
}

export const initialState: MapState = {
  map_styles: { layers: [] },
  map_overlays: [],
  occurrence_data: [],
  map_drawer: { open: false, overlays: false, baseMap: false },
};

export const getMapStyles = createAsyncThunk('map/getMapStyles', async () => {
  const mapStyles = await fetchMapStyles();
  return mapStyles;
});

export const getTileServerOverlays = createAsyncThunk(
  'map/getTileServerOverlays',
  async () => {
    const tileServerOverlays = await fetchTileServerOverlays();
    return tileServerOverlays;
  }
);

//Get occurrence results
export const getOccurrenceData = createAsyncThunk(
  'map/getSiteLocations',
  async (_: void, thunkAPI) => {
    const numberOfItemsPerResponse = 100;
    const response = await fetchGraphQlData(
      locationsQuery(0, numberOfItemsPerResponse)
    );
    var siteLocations = response.data.OccurrenceData.items;
    var hasMore = response.data.OccurrenceData.hasMore;
    var responseNumber = numberOfItemsPerResponse;
    thunkAPI.dispatch(updateOccurrence(siteLocations));
    while (hasMore === true) {
      const anotherResponse = await fetchGraphQlData(
        locationsQuery(responseNumber, numberOfItemsPerResponse)
      );
      const moreSiteLocations = anotherResponse.data.OccurrenceData.items;
      thunkAPI.dispatch(
        updateOccurrence([...siteLocations, ...moreSiteLocations])
      );
      siteLocations = [...siteLocations, ...moreSiteLocations];
      hasMore = anotherResponse.data.OccurrenceData.hasMore;
      responseNumber += numberOfItemsPerResponse;
    }
  }
);

export const mapSlice = createSlice({
  name: 'map',
  initialState,
  reducers: {
    updateOccurrence(state, action) {
      state.occurrence_data = action.payload;
    },
    drawerToggle(state) {
      const map_drawer = state.map_drawer;
      if (state.map_drawer.open === true) {
        map_drawer.open = false;
        map_drawer.overlays = false;
        map_drawer.baseMap = false;
      } else {
        map_drawer.open = true;
      }
    },

    drawerListToggle(state, action: PayloadAction<String>) {
      action.payload === 'overlays'
        ? (state.map_drawer.overlays = !state.map_drawer.overlays)
        : (state.map_drawer.baseMap = !state.map_drawer.baseMap);
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getMapStyles.fulfilled, (state, action) => {
        state.map_styles = action.payload;
      })
      .addCase(getTileServerOverlays.fulfilled, (state, action) => {
        state.map_overlays = action.payload;
      });
  },
});

export const { updateOccurrence, drawerToggle, drawerListToggle } =
  mapSlice.actions;
export default mapSlice.reducer;
