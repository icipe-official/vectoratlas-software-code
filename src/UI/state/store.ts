import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import rootReducer from './rootReducer';

export function makeStore() {
  return configureStore({
    reducer: rootReducer,
  });
}

const store = makeStore();

export type AppState = ReturnType<typeof store.getState>;
export const rootInitialState = store.getState() as AppState;

export type AppDispatch = typeof store.dispatch;

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  AppState,
  unknown,
  Action<string>
>;

export default store;
