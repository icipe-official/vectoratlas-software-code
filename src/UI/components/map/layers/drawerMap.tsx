import React from 'react';
import { useDispatch } from 'react-redux';
import { useAppSelector } from '../../../state/hooks';
import List from '@mui/material/List';
import { styled, useTheme } from '@mui/material/styles';
import Drawer from '@mui/material/Drawer';
import Divider from '@mui/material/Divider';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import ChevronLeftIcon from '@mui/icons-material/ChevronLeft';
import { DrawerList } from './drawerList';
import { Box } from '@mui/system';
import { drawerToggle } from '../../../state/mapSlice';

export default function DrawerMap() {
  const theme = useTheme();
  const dispatch = useDispatch();
  const drawerWidth = 240;
  const overlays = useAppSelector((state) =>
    state.map.map_overlays.filter((l: any) => l.sourceLayer !== 'world')
  );
  const baseMap = useAppSelector((state) =>
    state.map.map_overlays.filter((l: any) => l.sourceLayer === 'world')
  );
  const open = useAppSelector((state) => state.map.map_drawer.open);

  const handleDrawer = () => {
    dispatch(drawerToggle());
  };

  const openedMixin = (theme: any) => ({
    width: drawerWidth,
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
    overflowX: 'hidden',
    margin: '0px',
    marginTop: 18,
    height: '80%',
  });

  const closedMixin = (theme: any) => ({
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    overflowX: 'hidden',
    margin: '0px',
    marginTop: 18,
    height: '80%',
    width: `calc(${theme.spacing(7)} + 1px)`,
    [theme.breakpoints.up('sm')]: {
      width: `calc(${theme.spacing(8)} + 1px)`,
    },
  });

  const drawerHeaderSx = {
    display: 'flex',
    alignItems: 'center',
    justifyContent: open ? 'flex-end' : 'center',
    padding: 0,
    // necessary for content to be below app bar
    ...theme.mixins.toolbar,
  };

  const drawerSx = {
    width: drawerWidth,
    flexShrink: 0,
    whiteSpace: 'nowrap',
    boxSizing: 'border-box',
    ...(open && {
      ...openedMixin(theme),
      '& .MuiDrawer-paper': openedMixin(theme),
    }),
    ...(!open && {
      ...closedMixin(theme),
      '& .MuiDrawer-paper': closedMixin(theme),
    }),
  };

  return (
    <Drawer sx={drawerSx} variant="permanent" open={open} data-testid="drawer">
      <Box sx={drawerHeaderSx}>
        <IconButton data-testid="drawerToggle" onClick={handleDrawer}>
          {open === true ? (
            <ChevronLeftIcon data-testid="openDrawerChevron" />
          ) : (
            <MenuIcon />
          )}
        </IconButton>
      </Box>
      <List>
        <Divider />
        <DrawerList
          sectionTitle="Overlays"
          overlays={overlays}
          sectionFlag="overlays"
        />
        <Divider />
        <DrawerList
          sectionTitle="Base Map"
          overlays={baseMap}
          sectionFlag="baseMap"
        />
        <Divider />
      </List>
    </Drawer>
  );
}
