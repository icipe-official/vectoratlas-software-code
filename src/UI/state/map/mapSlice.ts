import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import {
  fetchGraphQlData,
  fetchMapStyles,
  fetchSpeciesList,
  fetchTileServerOverlays,
} from '../../api/api';
import { occurrenceQuery, fullOccurrenceQuery } from '../../api/queries';
import { unpackOverlays } from './mapSliceUtils';
import { VectorAtlasFilters } from '../state.types';
import { AppState } from '../store';

const countryList = [
  'Algeria',
  'Angola',
  'Benin',
  'Botswana',
  'Burkina Faso',
  'Burundi',
  'Cabo Verde',
  'Cameroon',
  'Central African Republic (CAR)',
  'Chad',
  'Comoros',
  'Congo, Democratic Republic of the',
  'Congo, Republic of the',
  'Cote dâ€™Ivoire',
  'Djibouti',
  'Egypt',
  'Equatorial Guinea',
  'Eritrea',
  'Eswatini',
  'Ethiopia',
  'Gabon',
  'Gambia',
  'Ghana',
  'Guinea',
  'Guinea-Bissau',
  'Kenya',
  'Lesotho',
  'Liberia',
  'Libya',
  'Madagascar',
  'Malawi',
  'Mali',
  'Mauritania',
  'Mauritius',
  'Morocco',
  'Mozambique',
  'Namibia',
  'Niger',
  'Nigeria',
  'Rwanda',
  'Sao Tome and Principe',
  'Senegal',
  'Seychelles',
  'Sierra Leone',
  'Somalia',
  'South Africa',
  'South Sudan',
  'Sudan',
  'Tanzania',
  'Togo',
  'Tunisia',
  'Uganda',
  'Zambia',
  'Zimbabwe',
];

const speciesList = [
  'coustani',
  'pauliani',
  'comorensis',
  'ovengensis',
  'cinctus',
  'bervoetsi',
  'njombiensis',
  'okuensis',
  'berghei',
  'mortiauxi',
  'funestus-like',
  'arabiensis',
  'culicifacies',
  'fuscivenosus',
  'quadriannulatus',
  'dthali',
  'domicola',
  'fontinalis',
  'squamosus',
  'hancocki',
  'faini',
  'maculipalpis',
  'funestus',
  'buxtoni',
  'dureni',
  'jebudensis',
  'rhodesiensis',
  'grassei',
  'paludis',
  'seretsei',
  'erythraeus',
  'eouzani',
  'notleyi',
  'gabonensis',
  'hargreavesi',
  'dancalicus',
  'somalicus',
  'longipalpis',
  'kingi',
  'vernus',
  'cameroni',
  'fontenillei',
  'hervyi',
  'brucei',
  'ranci',
  'vinckei',
  'rivulorum-like',
  'radama',
  'erepens',
  'confusus',
  'austenii',
  'letabensis',
  'cydippis',
  'seydeli',
  'deemingi',
  'carteri',
  'wellcomei',
  'christyi',
  'concolor',
  'keniensis',
  'ruarinus',
  'moucheti',
  'grenieri',
  'mascarensis',
  'barberellus',
  'stephensi',
  'gibbinsi',
  'lovettae',
  'flavicosta',
  'murphyi',
  'milloti',
  'azevedoi',
  'turkhudi',
  'griveaudi',
  'brohieri',
  'roubaudi',
  'mousinhoi',
  'daudi',
  'multicolor',
  'multicinctus',
  'caroni',
  'carnevalei',
  'swahilicus',
  'cinereus',
  'harperi',
  'aruni',
  'fuscicolor',
  'ardensis',
  'nili',
  'rufipes',
  'freetownensis',
  'machardyi',
  'obscurus',
  'hamoni',
  'ziemanni',
  'dualaensis',
  'vaneedeni',
  'rivulorum',
  'walravensi',
  'garnhami',
  'vanhoofi',
  'kosiensis',
  'demeilloni',
  'argenteolobatus',
  'schwetzi',
  'natalensis',
  'marshallii',
  'ethiopicus',
  'lacani',
  'bwambae',
  'pharoensis',
  'namibiensis',
  'implexus',
  'parensis',
  'merus',
  'distinctus',
  'pretoriensis',
  'hughi',
  'gambiae',
  'symesi',
  'listeri',
  'lounibosi',
  'caliginosus',
  'lloreti',
  'tenebrosus',
  'smithii',
  'wilsoni',
  'maliensis',
  'azaniae',
  'salbaii',
  'crypticus',
  'sergentii',
  'theileri',
  'amharicus',
  'millecampsi',
  'rageaui',
  'brunnipes',
  'brumpti',
  'leesoni',
  'tchekedii',
  'melas',
  'coluzzii',
  'cristipalpis',
  'rodhaini',
];

export function singularOutputs(filters: VectorAtlasFilters) {
  const updatedFilters = JSON.parse(JSON.stringify(filters));
  updatedFilters.country.value = filters.country.value[0];
  updatedFilters.species.value = filters.species.value[0];
  updatedFilters.season.value = filters.season.value[0];
  updatedFilters.control.value = filters.control.value[0];
  updatedFilters.isAdult.value = filters.isAdult.value[0];
  updatedFilters.isLarval.value = filters.isLarval.value[0];
  return updatedFilters;
}

export interface DetailedOccurrence {
  id: string;
  year_start: number;
  month_start: number;
  sample: {
    mossamp_tech_1: string;
  };
  recorded_species: {
    species: string;
  };
  reference: {
    author: string;
    year: number;
    citation: string;
  };
  bionomics: {
    adult_data: boolean;
    larval_site_data: boolean;
    season_given: string | null;
    season_calc: string | null;
  };
}

export interface MapState {
  map_styles: {
    layers: {
      name: string;
      colorChange: 'fill' | 'stroke';
      fillColor: number[];
      strokeColor: number[];
      strokeWidth: number;
      zIndex: number;
    }[];
  };
  map_overlays: {
    name: string;
    displayName: string;
    sourceLayer: string;
    sourceType: string;
    isVisible: boolean;
  }[];
  currentSearchID: string;
  occurrence_data: {
    items: [{}];
    total: number;
    hasMore: boolean;
  }[];
  map_drawer: {
    open: boolean;
    overlays: boolean;
    baseMap: boolean;
    filters: boolean;
    download: boolean;
  };
  filters: VectorAtlasFilters;
  filterValues: {
    country: string[];
    species: string[];
  };
  species_list: { series: string; color: number[] }[];
  selectedIds: string[];
  selectedData: DetailedOccurrence[];
}

export const initialState: () => MapState = () => ({
  map_styles: { layers: [] },
  map_overlays: [],
  occurrence_data: [],
  currentSearchID: '',
  map_drawer: {
    open: false,
    overlays: false,
    baseMap: false,
    filters: false,
    download: false,
  },
  species_list: [],
  filters: {
    country: { value: [] },
    species: { value: [] },
    isLarval: { value: [] },
    isAdult: { value: [] },
    control: { value: [] },
    season: { value: [] },
    timeRange: {
      value: {
        start: null,
        end: null,
      },
    },
  },
  filterValues: {
    country: countryList,
    species: speciesList,
  },
  selectedIds: [],
  selectedData: [],
});

export const getMapStyles = createAsyncThunk('map/getMapStyles', async () => {
  const mapStyles = await fetchMapStyles();
  return mapStyles;
});

export const getTileServerOverlays = createAsyncThunk(
  'map/getTileServerOverlays',
  async () => {
    const tileServerOverlays = await fetchTileServerOverlays();
    return tileServerOverlays;
  }
);

export const getSpeciesList = createAsyncThunk(
  'map/getSpeciesList',
  async () => {
    const speciesList = await fetchSpeciesList();
    return speciesList.data;
  }
);

//Get occurrence results
export const getOccurrenceData = createAsyncThunk(
  'map/getOccurrenceData',
  async (filters: MapState['filters'], thunkAPI) => {
    const numberOfItemsPerResponse = 100;
    const response = await fetchGraphQlData(
      occurrenceQuery(0, numberOfItemsPerResponse, singularOutputs(filters))
    );

    var siteLocations = response.data.OccurrenceData.items;
    var hasMore = response.data.OccurrenceData.hasMore;
    var responseNumber = numberOfItemsPerResponse;

    const searchID = 'id' + Math.random().toString(16).slice(2);
    thunkAPI.dispatch(startNewSearch(searchID));
    thunkAPI.dispatch(updateOccurrence({ data: siteLocations, searchID }));
    while (hasMore === true) {
      const anotherResponse = await fetchGraphQlData(
        occurrenceQuery(
          responseNumber,
          numberOfItemsPerResponse,
          singularOutputs(filters)
        )
      );
      const moreSiteLocations = anotherResponse.data.OccurrenceData.items;
      thunkAPI.dispatch(
        updateOccurrence({
          data: [...siteLocations, ...moreSiteLocations],
          searchID,
        })
      );
      siteLocations = [...siteLocations, ...moreSiteLocations];
      hasMore = anotherResponse.data.OccurrenceData.hasMore;
      responseNumber += numberOfItemsPerResponse;
    }
  }
);

export const getFullOccurrenceData = createAsyncThunk(
  'map/getFullOccurrenceData',
  async (_, thunkAPI) => {
    const selectedIds = (thunkAPI.getState() as AppState).map.selectedIds;
    const response = await fetchGraphQlData(fullOccurrenceQuery(selectedIds));
    const data = response.data.FullOccurrenceData;
    thunkAPI.dispatch(updateSelectedData(data));
  }
);

export const mapSlice = createSlice({
  name: 'map',
  initialState: initialState(),
  reducers: {
    setSelectedIds(state, action) {
      state.selectedIds = action.payload;
    },
    startNewSearch(state, action) {
      state.currentSearchID = action.payload;
    },
    updateSelectedData(state, action) {
      state.selectedData = action.payload;
    },
    updateOccurrence(state, action) {
      if (action.payload.searchID === state.currentSearchID) {
        state.occurrence_data = action.payload.data;
      }
    },
    drawerToggle(state) {
      const map_drawer = state.map_drawer;
      if (state.map_drawer.open === true) {
        map_drawer.open = false;
        map_drawer.overlays = false;
        map_drawer.baseMap = false;
        map_drawer.filters = false;
        map_drawer.download = false;
      } else {
        map_drawer.open = true;
      }
    },
    drawerListToggle(state, action: PayloadAction<String>) {
      action.payload === 'overlays'
        ? (state.map_drawer.overlays = !state.map_drawer.overlays)
        : action.payload === 'baseMap'
        ? (state.map_drawer.baseMap = !state.map_drawer.baseMap)
        : action.payload === 'download'
        ? (state.map_drawer.download = !state.map_drawer.download)
        : (state.map_drawer.filters = !state.map_drawer.filters);
    },
    layerToggle(state, action: PayloadAction<String>) {
      const overlayToToggle = state.map_overlays.find(
        (l: any) => l.name === action.payload
      );
      if (!overlayToToggle) {
        return;
      }
      overlayToToggle.isVisible = !overlayToToggle.isVisible;
    },
    filterHandler(state: any, action) {
      state.filters[action.payload.filterName].value =
        action.payload.filterOptions;
    },
    updateMapLayerColour(state, action) {
      const matchingLayer = state.map_styles.layers.find(
        (l) => l.name === action.payload.name
      );
      if (matchingLayer) {
        if (matchingLayer.colorChange === 'fill') {
          matchingLayer.fillColor = action.payload.color;
        } else {
          matchingLayer.strokeColor = action.payload.color;
        }
      }
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getMapStyles.fulfilled, (state, action) => {
        state.map_styles = action.payload;
      })
      .addCase(getTileServerOverlays.fulfilled, (state, action) => {
        state.map_overlays = unpackOverlays(action.payload);
      })
      .addCase(getSpeciesList.fulfilled, (state, action) => {
        state.species_list = action.payload;
      });
  },
});

export const {
  updateOccurrence,
  updateSelectedData,
  drawerToggle,
  drawerListToggle,
  layerToggle,
  filterHandler,
  startNewSearch,
  updateMapLayerColour,
  setSelectedIds,
} = mapSlice.actions;
export default mapSlice.reducer;
