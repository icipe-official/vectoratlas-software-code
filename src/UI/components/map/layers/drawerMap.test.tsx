import React from 'react';
import { initialState } from '../../../state/mapSlice';
import { AppState } from '../../../state/store';
import { render } from '../../../test_config/render';
import { fireEvent, screen } from '@testing-library/dom';
import { DrawerMap } from './drawerMap';

describe(DrawerMap.name, () => {
  it('renders drawer', () => {
    const state: Partial<AppState> = {
      map: {
        ...initialState,
        map_overlays: [
          {
            name: 'an_gambiae',
            sourceLayer: 'overlays',
            sourceType: 'raster',
            overlays: [],
          },
          {
            name: 'world',
            sourceLayer: '',
            sourceType: 'vector',
            overlays: [
              { name: 'countries' },
              { name: 'lakes_reservoirs' },
              { name: 'land' },
              { name: 'oceans' },
              { name: 'rivers_lakes' },
            ],
          },
        ],
      },
    };

    render(<DrawerMap />, state);

    expect(screen.getByTestId('drawer')).toBeVisible();
  });

  it('the state of the drawer can be toggled when menu button and chevron left are clicked', () => {
    let testState;
    testState = { map: JSON.parse(JSON.stringify(initialState)) };
    testState.map.map_drawer = { open: true, overlays: true, baseMap: true };

    const { store } = render(<DrawerMap />, testState);
    fireEvent.click(screen.getByTestId('drawerToggle'));

    const actions = store.getActions();
    expect(actions).toHaveLength(1);
    expect(actions[0]).toEqual({ type: 'map/drawerToggle' });
  });
});
