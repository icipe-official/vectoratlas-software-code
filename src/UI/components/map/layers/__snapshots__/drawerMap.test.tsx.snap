// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerMap renders a filter download 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1tx2ity-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  Download Filtered Data
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`DrawerMap renders a filter download button 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1tx2ity-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  Download Filtered Data
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`DrawerMap renders filter list 1`] = `
<div>
  Filter list mock 
  {"sectionTitle":"Filters","sectionFlag":"filters"}
</div>
`;
